plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -782,6 +997,147 @@ int pack_emf_payload(struct flb_cloudwatch *ctx,
+                        if(entity->attributes->name_source == NULL) {
+                            entity->filter_count++;
+                        } else {
+                            flb_free(entity->attributes->name_source);
+                        }
+                        entity->attributes->name_source = flb_strndup(kube_val.via.str.ptr, kube_val.via.str.size);
+                    } else if(strncmp(kube_key.via.str.ptr, "aws_entity_platform", kube_key.via.str.size) == 0) {
+                        if(entity->attributes->platform_type == NULL) {
+                            entity->filter_count++;
+                        } else {
+                            flb_free(entity->attributes->platform_type);
+                        }
+                        entity->attributes->platform_type = flb_strndup(kube_val.via.str.ptr, kube_val.via.str.size);
+                    }
+                }
+            }
+        }
+        if(strncmp(key.via.str.ptr, "aws_entity_ec2_instance_id",key.via.str.size ) == 0 ) {
+            if(entity->attributes->instance_id == NULL) {
+                entity->root_filter_count++;
+            } else {
+                flb_free(entity->attributes->instance_id);
+            }
+            entity->attributes->instance_id = flb_strndup(val.via.str.ptr, val.via.str.size);
+        }
+    }
+    if(entity->key_attributes->name == NULL && entity->attributes->name_source == NULL &&entity->attributes->workload != NULL) {
+        entity->key_attributes->name = flb_strndup(entity->attributes->workload, strlen(entity->attributes->workload));
+        entity->attributes->name_source = flb_strndup("K8sWorkload", 11);
+    }
+    if(entity->key_attributes->environment == NULL) {
+        entity->key_attributes->environment = find_fallback_environment(ctx, entity);
+    }
+}
+
+void update_or_create_entity(struct flb_cloudwatch *ctx, struct log_stream *stream, const msgpack_object map) {
+        if(stream->entity == NULL) {
+            stream->entity = flb_malloc(sizeof(entity));
+            if (stream->entity == NULL) {
+                return;
+            }
+            memset(stream->entity, 0, sizeof(entity));
+
+            stream->entity->key_attributes = flb_malloc(sizeof(entity_key_attributes));
+            if (stream->entity->key_attributes == NULL) {
+                return;
+            }
+            memset(stream->entity->key_attributes, 0, sizeof(entity_key_attributes));
+
+            stream->entity->attributes = flb_malloc(sizeof(entity_attributes));
+            if (stream->entity->attributes == NULL) {
+                return;
+            }
+            memset(stream->entity->attributes, 0, sizeof(entity_attributes));
+            stream->entity->filter_count = 0;
+            stream->entity->root_filter_count = 0;
+
+            parse_entity(ctx,stream->entity,map, map.via.map.size);
```

Comment:
Nit: This looks identical to line 1134; it could probably be factored out of this `if` statement?

Reply:
addressed in new commit